commands/command_migrate_export.go

```diff
@@ -0,0 +1,155 @@
+package commands
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/git-lfs/git-lfs/errors"
+	"github.com/git-lfs/git-lfs/filepathfilter"
+	"github.com/git-lfs/git-lfs/git"
+	"github.com/git-lfs/git-lfs/git/githistory"
+	"github.com/git-lfs/git-lfs/git/odb"
+	"github.com/git-lfs/git-lfs/lfs"
+	"github.com/git-lfs/git-lfs/tasklog"
+	"github.com/git-lfs/git-lfs/tools"
+	"github.com/spf13/cobra"
+)
+
+func migrateExportCommand(cmd *cobra.Command, args []string) {
+	l := tasklog.NewLogger(os.Stderr)
+	defer l.Close()
+
+	db, err := getObjectDatabase()
+	if err != nil {
+		ExitWithError(err)
+	}
+	defer db.Close()
+
+	rewriter := getHistoryRewriter(cmd, db, l)
+
+	filter := rewriter.Filter()
+	if len(filter.Include()) <= 0 {
+		ExitWithError(errors.Errorf("fatal: one or more files must be specified with --include"))
+	}
+
+	tracked := trackedFromExportFilter(filter)
+	gitfilter := lfs.NewGitFilter(cfg)
+
+	opts := &githistory.RewriteOptions{
+		Verbose:           migrateVerbose,
+		ObjectMapFilePath: objectMapFilePath,
+		BlobFn: func(path string, b *odb.Blob) (*odb.Blob, error) {
+			if filepath.Base(path) == ".gitattributes" {
+				return b, nil
+			}
+
+			var buf bytes.Buffer
+
+			if _, err := smudge(gitfilter, &buf, b.Contents, path, false, rewriter.Filter()); err != nil {
+				return nil, err
+			}
+
+			return &odb.Blob{
+				Contents: &buf, Size: int64(buf.Len()),
+			}, nil
+		},
+
+		TreeCallbackFn: func(path string, t *odb.Tree) (*odb.Tree, error) {
+			if path != "/" {
+				// Ignore non-root trees.
+				return t, nil
+			}
+
+			ours := tracked
+			theirs, err := trackedFromAttrs(db, t)
+			if err != nil {
+				return nil, err
+			}
+
+			// Create a blob of the attributes that are optionally
+			// present in the "t" tree's .gitattributes blob, and
+			// union in the patterns that we've tracked.
+			//
+			// Perform this Union() operation each time we visit a
+			// root tree such that if the underlying .gitattributes
+			// is present and has a diff between commits in the
+			// range of commits to migrate, those changes are
+			// preserved.
+			blob, err := trackedToBlob(db, theirs.Clone().Union(ours))
+			if err != nil {
+				return nil, err
+			}
+
+			// Finally, return a copy of the tree "t" that has the
+			// new .gitattributes file included/replaced.
+			return t.Merge(&odb.TreeEntry{
+				Name:     ".gitattributes",
+				Filemode: 0100644,
+				Oid:      blob,
+			}), nil
+		},
+
+		UpdateRefs: true,
+	}
+
+	requireInRepo()
+
+	opts, err = rewriteOptions(args, opts, l)
+	if err != nil {
+		ExitWithError(err)
+	}
+
+	// If we have a valid remote, pre-download all objects using the Transfer Queue
+	if remoteURL := getAPIClient().Endpoints.RemoteEndpoint("download", cfg.Remote()).Url; remoteURL != "" {
```

Comment:
Nice. What do you think about allowing `[--remote=<remote>]` as an optional argument in this command, to specify where clients would like to get their objects from?

We could even take it a step further and allow them to specify multiple remotes. Package `tq` today doesn't know how to interact with multiple remotes at once (and I'm inclined not to teach it how to do so), but we could work around this by looping through each given remote. Maybe something like:

```go
var wg sync.WaitGroup

for _, remote := range remotes {
        wg.Add(1)
        go downloadObjects(remotes, wg)
}

wg.Wait()
```

Reply:
I think that's a good idea! We can just default to `cfg.Remote()` like we do here if the argument isn't provided.

Do we want to support multiple remotes for this? What would the use cases for that look like? I'm concerned about potential conflicts in object downloads, for example if one provided remote points to a fork of another repo, which is also provided as a remote.

Reply:
I think that supporting just the one remote is fine for now; I can see this becoming somewhat complicated because the transfer queue really isn't designed for multi-remote operations.

I think that there are a few approaches we can take in the longer term, but for now, I think that supporting one remote via `--remote` is totally good. If a user is (1) the bleeding-edge version of Git LFS, and (2) wants to do a multi-remote export, they can always use `git lfs fetch && git lfs migrate export` instead.

Longer-term, I think we could do one of the following:

1. Teach package `tq` how to handle multiple remotes. This would provide a consistent interface for using the transfer queue, but will be only used at this singular call-site, and seems too complicated for too little benefit.

2. Keep many transfer queues alive, and round-robin the objects through them. This has the benefit of simplicity and no requirements for change inside of package `tq`, and the external call-sites should be fairly straightforward, too. If an earlier transfer queue has the same object as a later one, it doesn't really matter which one answers first. This has the drawback of having to wait for one transfer queue to cycle before the next one in line can determine which objects they need to get.

I think that we could theoretically get _more_ efficient than that, but I bet that that's an OK starting point for now.